gobblin-api/src/main/java/gobblin/source/extractor/WatermarkInterval.java

```diff
@@ -0,0 +1,91 @@
+package gobblin.source.extractor;
+
+import gobblin.fork.CopyNotSupportedException;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.codehaus.jackson.map.ObjectMapper;
+
+import com.google.common.base.Preconditions;
+
+public class WatermarkInterval implements Writable {
+
+  private Watermark lowWatermark;
+  private Watermark expectedHighWatermark;
+  private Watermark actualHighWatermark;
+
+  private ObjectMapper objectMapper;
+
+  // Needed for the Writable interface
+  public WatermarkInterval() {
+  }
+
+  private WatermarkInterval(Watermark lowWatermark, Watermark expectedHighWatermark) throws CopyNotSupportedException {
+    this.lowWatermark = lowWatermark;
+    this.expectedHighWatermark = expectedHighWatermark;
+    this.actualHighWatermark = lowWatermark.copy();
+
+    this.objectMapper = new ObjectMapper();
+  }
+
+  public static class Builder {
+
+    private Watermark lowWatermark;
+    private Watermark expectedHighWatermark;
+
+    public Builder withLowWatermark(Watermark lowWatermark) {
+      this.lowWatermark = lowWatermark;
+      return this;
+    }
+
+    public Builder withExpectedHighWatermark(Watermark expectedHighWatermark) {
+      this.expectedHighWatermark = expectedHighWatermark;
+      return this;
+    }
+
+    public WatermarkInterval build() throws CopyNotSupportedException {
+      Preconditions.checkNotNull(this.lowWatermark, "Must specify a low watermark");
+      Preconditions.checkNotNull(this.expectedHighWatermark, "Must specify an expected high watermark");
+
+      return new WatermarkInterval(this.lowWatermark, this.expectedHighWatermark);
+    }
+  }
+
+  public void increment(Object record) {
+    this.actualHighWatermark.increment(record);
+  }
+
+  public boolean isTargetReached() {
+    return this.actualHighWatermark.equals(this.expectedHighWatermark);
```

Comment:
Hmm, I'm not sure if we should make this an IllegalState. While I can't think of a concrete example, there may be users who want to pull data past the expected high watermark. For example, if the expected high watermark can't really be known until runtime, or if the `WorkUnit` pulls data until the maximum offset is reached.